tools/dep_tree_resolver/go_deps.go

```diff
@@ -0,0 +1,413 @@
+	return dependencies
+}
+
+func isCirccularDependency(module *Module, depPath *[]Module) bool {
+	for _, seenModule := range *depPath {
+		if (*module).Path == seenModule.Path {
+			return true
+		}
+	}
+
+	return false
+}
+
+func resolveRecursive(
+	module *Module,
+	level int,
+	rootDepPath []Module,
+	depFlatGraph []ModuleDep,
+	actualModuleVersions map[string]Module,
+) (*DependencyTree, error) {
+
+	modPath := module.Path
+	if Debug {
+		fmt.Printf("Resolving dependency (Depth: %d) %s\n", level, modPath)
+	}
+
+	actualModule, ok := actualModuleVersions[modPath]
+	if !ok {
+		return nil,
+			fmt.Errorf("ERROR! Could not find actual module mapping for %s (level %d). Path: %+v", modPath, level, rootDepPath)
+	}
+
+	modDeps := getModuleDependencies(depFlatGraph, actualModule)
+
+	if Debug {
+		fmt.Printf("- Found %d dep(s) for module %s\n", len(modDeps), actualModule)
+	}
+
+	dependencies := []DependencyTree{}
+	for _, depModule := range modDeps {
+		// Circular dependencies are ignored once we encounter a loop
+		if isCirccularDependency(&depModule, &rootDepPath) {
+			if Debug {
+				fmt.Printf("WARN: Circular dependency breakout for %s\n", depModule.Path)
+			}
+			continue
+		}
+
+		depPath := append(rootDepPath, *module)
+		depSubTree, err := resolveRecursive(&depModule, level+1, depPath, depFlatGraph, actualModuleVersions)
+		if err != nil {
+			return nil, err
+		}
+
+		if Debug {
+			fmt.Printf("- Appending dep tree of %s to %s (%s)\n", depModule.Path, modPath, actualModule.Path)
+		}
+
+		dependencies = append(dependencies, *depSubTree)
+	}
+
+	return &DependencyTree{
+		Mod:          actualModule,
+		Dependencies: dependencies,
+	}, nil
+}
+
+func recomputeDependencyTree(
+	rootModule *Module,
+	depFlatGraph []ModuleDep,
+	actualModuleVersions map[string]Module,
+) (*DependencyTree, error) {
+	// Main root node
+	depTree := DependencyTree{
+		Mod: Module{
+			Path:    "root",
+			Version: "root",
+		},
+	}
+
+	// Some deps have circular dependencies so we need to break out when
+	// we encounter them
+	rootDepPath := []Module{*rootModule}
+
+	// Queue up the main root node deps and recurseviley resolve
+	rootDeps := getModuleDependencies(depFlatGraph, *rootModule)
+	for _, rootDep := range rootDeps {
+		module := rootDep
+		if Debug {
+			fmt.Printf("Resolving root dependency %s\n", module)
+		}
+		depPath := append(rootDepPath, module)
+		depSubtree, err := resolveRecursive(&module, 0, depPath, depFlatGraph, actualModuleVersions)
+		if err != nil {
+			return nil, err
+		}
+		depTree.Dependencies = append(depTree.Dependencies, *depSubtree)
+	}
+
+	return &depTree, nil
+}
+
+// TODO: Actually use `skipDuplicates` value
+func printDepTree(buf *bufio.Writer, depTree *DependencyTree, level int, skipDuplicates bool) {
+	for idx := 0; idx < level; idx++ {
+		io.WriteString(buf, "    ")
+	}
+
+	io.WriteString(buf, fmt.Sprintf("- %s@%s\n", (*depTree).Mod.Path, (*depTree).Mod.Version))
+	for _, depSubtree := range (*depTree).Dependencies {
```

Comment:
The deref for this particular line makes sense but passing around structs I think would:
- Slow down the already slow code
- Make it jump exponentially in memory usage as it would use copies of same structs (and we have a very large tree to work with)
- Would possibly cause unintended issues with modifying a module object within a method only to find out that the parent caller didn't have its sent arg changed as expected